### ISSUE TITLE
hatches/patterns don't work in vcs 2.0.beta

### ISSUE THREAD

**Issue by jypeter** [UV-CDAT/uvcdat#541]
The test script pasted at the end is supposed to plot a map with hatches, and a canvas with some test patterns

 * It works correctly with 1.5.1 (though I'm getting strange warning messages that I did not get in earlier versions)
![test_fillarea](https://cloud.githubusercontent.com/assets/2785573/3840372/99cb17b8-1e1c-11e4-8db1-fb8070f31cc6.png)
![test_fillarea_list](https://cloud.githubusercontent.com/assets/2785573/3840375/a6aa3540-1e1c-11e4-8a78-b24d3b3cbf8d.png)

 * In 2.0.beta, after typing *x.interact()* (I did not know about this one, but a Warning told me to use it), I get some *Nope no array[0]* messages and the display is not exactly what I expected. I have attached a screen copy below, since the generated png files are empty

![hatch_pb_2 0 beta](https://cloud.githubusercontent.com/assets/2785573/3840458/e425a688-1e1d-11e4-942f-054d7ba6cf22.png)

**The test script I used**

```
#!/usr/bin/env python

# J-Y Peterschmitt - LSCE - 09/2011 - pmip2web@lsce.ipsl.fr

# Test the use of hatches and patterns in the isofill
# and fill area graphics methods

# Import some standard modules
import sys, os
from os import path

# Import what we need from CDAT
import cdms2
import vcs

# Some data we can plot from the 'sample_data' directory
# supplied with CDAT
data_dir = path.join(sys.prefix, 'sample_data')
data_file = 'tas_ccsr-95a_1979.01-1979.12.nc'
var_name = 'tas'

# Zone that we want to plot
#
# NOTE: the (latmin, latmax, lonmin, lonmax) information HAS TO be the
# same in the variable, the 'isof' isofill method and the 2 'cont_*'
# continents plotting methods! Otherwise, the data will not match the
# continents that are plotted over it...
(latmin, latmax, lonmin, lonmax) = (-90, 90, -180, 180)

# Use black on white continents (nicer with black and white plots) i.e
# we plot a 'large' white continent outline over the data, and then a
# smaller 'black' continent outline
bw_cont = False
#bw_cont = True

# Read one time step (the first one) from the data file
# and explicitely specify the lat/lon range we need. cdms2
# will retrieve the data in the order we want, regardless of the way
# it is stored in the data file
f = cdms2.open(path.join(data_dir, data_file))
v = f(var_name, time=slice(0, 1), latitude=(latmin, latmax),
      longitude=(lonmin, lonmax, 'co'), squeeze=1)
f.close()

# Initialize the graphics canvas
x = vcs.init()

# Create the isofill method
isof = x.createisofill('test_hatch')
isof.datawc(latmin, latmax, lonmin, lonmax)
isof.levels = [220, 240, 260, 280, 300, 320]
isof.fillareastyle = 'hatch'
isof.fillareacolors = [241, 241, 241, 241, 241] # All black
#isof.fillareacolors = [243, 248, 254, 255, 241] # Colors
isof.fillareaindices = [20, 9, 2, 7, 3]

# Define some graphics methods for plotting black on white continents
if bw_cont:
    cont_black = x.createcontinents('black')
    cont_black.datawc(latmin, latmax, lonmin, lonmax)
    cont_black.linecolor = 241
    cont_black.linewidth = 2
    cont_white = x.createcontinents('white')
    cont_white.datawc(latmin, latmax, lonmin, lonmax)
    cont_white.linecolor = 240
    cont_white.linewidth = 6

    cont_type = 0 # Do not plot the default continents
else:
    cont_type = 1

# Plot the test data
#
# We have to make sure the data and the continents are plotted at the
# same place ('data' area) on the canvas, by using the same template!
# It's even better if we can use for the continents a template that
# will only plot the data area (the priority of the other elements of
# the canvas is set to zero)
tpl = x.createtemplate('tpl', 'default')
x.plot(tpl, isof, v, continents=cont_type)
if bw_cont:
    tpl_data = x.createtemplate('tpl_data', 'default_dud') # plots only data area
    x.plot(tpl_data, cont_white)
    x.plot(tpl_data, cont_black)

# Create a test plot for listing all the hatches and patterns
style_list = []
index_list = []
col_cycle = [243, 248, 254, 241, 255]
nb_cols = len(col_cycle)
color_list = []
x_list = []
y_list = []
txt_x_list = []
txt_y_list = []
txt_str_list = []

shear_x = .05
for j, style in enumerate(['hatch', 'pattern']):
    slide_y = j * .4
    for i in range(20):
        slide_x = i * 0.04
        x1, y1 = (.05 + slide_x, .25 + slide_y)
        x2, y2 = (.08 + slide_x, .45 + slide_y)

        # Add (sheared) rectangles to the list of positions
        # NOTE: no need to close the fill area. Giving 4 vertices
        #       for getting a filled rectangle is enough
        x_list.append([x1, x2, x2 + shear_x, x1 + shear_x])
        y_list.append([y1, y1, y2, y2])

        style_list.append(style)
        # Hatches/Patterns indices have to be in 1-20 range
        index_list.append(i+1)
        col_idx = col_cycle[i % nb_cols]
        color_list.append(col_idx)

        # Annotations
        txt_x_list.append(x1 + 0.015)
        txt_y_list.append(y1 - 0.015)
        txt_str_list.append('%s = %i  -  Color = %i' %
                             (style, i+1, col_idx))

# Create the fill area and the text annotations
fill_test = x.createfillarea('fill_test')
fill_test.style = style_list
fill_test.index = index_list
fill_test.color = color_list
fill_test.x = x_list
fill_test.y = y_list

fill_info = x.createtext('fill_info')
fill_info.angle = 45
fill_info.height = 12
fill_info.color = 241 # Black
fill_info.string = txt_str_list
fill_info.x = txt_x_list
fill_info.y = txt_y_list

# Create a title
plot_title = x.createtext('plot_title')
plot_title.height = 40
plot_title.string = ['Testing hatches and patterns in VCS/CDAT']
plot_title.x = [.01]
plot_title.y = [.9]

# Initialize and use a second graphics canvas
y = vcs.init()
y.plot(plot_title)
y.plot(fill_test)
y.plot(fill_info)

# Save the plots
x.pdf('test_fillarea')
x.png('test_fillarea')
y.pdf('test_fillarea_list')
y.png('test_fillarea_list')

# Note: depending on the version of CDAT, text may not resize
#       correctly when creating a bigger png
#x.png('test_fillarea_big', width=3*11, height=3*8)
#y.png('test_fillarea_list_big', width=3*11, height=3*8)

# The end

```

**Comment by doutriaux1** [comment]
@aashish24 @dlonie any progress on this?

**Comment by aashish24** [comment]
@doutriaux1 @jypeter I have to remember what was the bug or how much @doutriaux1 had developed the code. I doubt that we can get this one in 2.2 (other higher priorities for now) but we can try.

**Comment by doutriaux1** [comment]
@aashish24 I think most of the code is already in, my guess is that my test png were too big. If you could explain me how to pass a VTK-made pattern rather a png as texture I think we would be set.

**Comment by doutriaux1** [comment]
@dlonie @aashish24 I guess we're moving this to 2.3...

**Comment by aashish24** [comment]
@dlonie is looking into this but I don't think we can make into 2.2

**Comment by dlonie** [comment]
@aashish24 @doutriaux1 Should we hold this off until the vtk/vcs layer rewrite? It looks like significant effort that will just be replaced in the near future.

**Comment by jypeter** [comment]
It's not a feature I use often, because I tend to forget about it, so it can wait a bit more, if need be

Hey, @chaosphere2112 , that's another example you can use for the gallery once the problems are fixed! :)

**Comment by aashish24** [comment]
@dlonie I won't suggest that. Can you describe the problem you are running into? I was hoping that it should not be a lot of work but I haven't look at the code. I can see some complications.

**Comment by chaosphere2112** [comment]
@jypeter Added to the list :+1:

**Comment by dlonie** [comment]
Here's where I stopped:

        if isinstance(mapper,list):
          ## This is the sport to add patterns
          #act.GetMapper().ScalarVisibilityOff()
          #act.SetTexture(mapper[1])
          pass

If `mapper` is a list in the plot2d megafunction, then somehow that means a hatch/pattern is needed. This is hacky, undocumented, and raises a lot of red flags. There's way too much going on here, and any changes I make are likely to break a dozen other things.

We should really just leave this code alone until we can rewrite in to a cleaner, better documented format where it's clear how each task is carried out.

This is going to tricky to implement and will need a custom override for the GL2PS exporter to get this working completely (GL2PS can't handle textures, I'll have to manually inject them into the image). I'd rather just get it right the first time and not have to deal with the fallout that will come from modifying this function.

**Comment by aashish24** [comment]
@dlonie I wouldn't worry about the cleaness of the code as far as we can write a method that applies the texture to the data layer. Once we have that function, we can always use that when we re-structure. Let's talk more about it today. May be I am missing sometihng.

**Comment by dlonie** [comment]
Talked with @aashish24, and I'll add a function that applies a texture to an actor and add a handler to the vector graphics exporter to copy the texture data over.

This will leave the following tasks to the VCS team:
- Produce textures to create the desired patterns.
- Set appropriate texture coordinates on the polydata points.
- Integrate the texture setup code into the VTKPlots.py.

**Comment by dlonie** [comment]
This is going to requires some C++ extensions to VTK to add a specialized GL2PSExporter subclass (Can't override the proper virtuals from Python). Do we have an existing C++ layer that I can add to, or should I add a new subproject?

**Comment by aashish24** [comment]
No existing C++ layer but in VCS there is some C code. May be within VCS, we can create a sub-project for it?

**Comment by dlonie** [comment]
@doutriaux1 How are these textures/patterns accessed in vcs?

**Comment by dlonie** [comment]
@doutriaux1 e.g. are there png files that have the patterns, or a bitmask representation somewhere?

**Comment by doutriaux1** [comment]
@dlonie I used to generate patterns via cairo. If you get it to work with any png it would be ok. Best way would be to pass some vtk generated pattern.

**Comment by doutriaux1** [comment]
@dlonie this is post release though there are still issues to be solved for release

**Comment by dlonie** [comment]
@sankhesh Grab me before starting on this one, I have some ideas on how we can accomplish this using pure python VTK.

**Comment by durack1** [comment]
@dlonie @sankhesh +100 - there are a heap of folks that are eagerly awaiting this updated functionality!!

**Comment by doutriaux1** [comment]
@aashish24 @sankhesh we are approaching ONE YEAR on this bug! This is very VERY HIGH priority! I would say top of @sankhesh list! @williams13 what do you think? There's is not a single week without a user requesting this.

**Comment by sankhesh** [comment]
@doutriaux1  I'll take a look at it today.

**Comment by aashish24** [comment]
@doutriaux1 we never set this one at the top in our priority list. We should mark issues with (high  and low; already existing labels). I would suggest we discuss priorities in our two meetings specially when you hear folks asking about it (since they didn't report it to the mailing list).

**Comment by doutriaux1** [comment]
@aashish24 I have been mentioning it to pretty much every meeting ever since the issue was reported :wink:

**Comment by aashish24** [comment]
@doutriaux1 it wasn't mentioned in the last few meetings (others can confirm). Also, if you look back at our meeting documents, you won't find it there in the last 4 meetings.

**Comment by doutriaux1** [comment]
Hum... Pretty sure I mention it at every wednesday/kitware meeting...

**Comment by doutriaux1** [comment]
@sankhesh place holder used to be here: 

https://github.com/UV-CDAT/uvcdat/blob/45488cb2a93fc3c0b1052e8ea26e404766038dc9/Packages/vcs/Lib/VTKPlots.py#L1061-1066
https://github.com/UV-CDAT/uvcdat/blob/45488cb2a93fc3c0b1052e8ea26e404766038dc9/Packages/vcs/Lib/VTKPlots.py#L1218-1222

**Comment by aashish24** [comment]
@doutriaux1 sorry no you didnt' mention :cry:

**Comment by doutriaux1** [comment]
anyhow it's a top priority now

**Comment by bonfils2** [comment]
please, fix the hatches as soon as possible! This is critical for me! Thanks!

**Comment by durack1** [comment]
@aashish24 @sankhesh +5 for me too..

**Comment by sankhesh** [comment]
@doutriaux1

I see that the hatches/patterns are accessed using an index in range from 1 to 20. Where is the list of patterns that this index points to?

**Comment by aashish24** [comment]
@sankhesh yup, this is a top priority now since we took care of others as of now. Ping @doutriaux1 or me as much as you need.

**Comment by dlonie** [comment]
@sankhesh I had a similar question on March 12 above. I don't think they care about matching the old version's patterns, just that there is something.

**Comment by aashish24** [comment]
Yup, first pass would be to get something working (just so that we have patterns). In a second pass we can try to match the patterns (and that should be a easier task). 

- Aashish

**Comment by chaosphere2112** [comment]
There's the example at the top of this thread that shows all of the patterns, though I don't know of any other documentation that lists them.

**Comment by doutriaux1** [comment]
@sankhesh yes as @chaosphere2112 the pattern are shown on the pix above. Code to generate them in cairo are at: https://github.com/UV-CDAT/uvcdat/blob/release/Packages/vcs_legacy/Src/xgks/cairo/meta2cairo.c#L616-L1073
Ideally it would be nice to be able to add a "custom" type and pass a png file for it. But that is not necessary.

**Comment by doutriaux1** [comment]
@sankhesh also pattern and hatch are essentially the same except pattern are black and white and hatches are colored instead of black.

**Comment by jypeter** [comment]
They are probably listed in an old xgks documentation... Try to emulate the matplotlib patterns! ;-) 
-- 
JYP @ SGS 2

**Comment by sankhesh** [comment]
Sounds good. I'll post some images here once I have something.

**Comment by sankhesh** [comment]
**UPDATE:**

I have this working partially with isofill plots. 

![test_fillarea](https://cloud.githubusercontent.com/assets/936354/9117386/cef3cad6-3c37-11e5-8e7b-4db641c778b3.png)

I still have some issues that I am looking at.

* Figure out why only a partial area is filled with pattern
* Understand difference in contourLevels and contour values
* Add support for different patterns and hatches
* Provide the same infrastructure for other fillareas
* Add support for patterns/hatches in the plot legend

**Comment by sankhesh** [comment]
UPDATE:
![test_fillarea](https://cloud.githubusercontent.com/assets/936354/9347954/6dfb09b2-45fe-11e5-81de-7183a4affef1.png)

Very close now. 
Remaining issues:

* Currently only 6 patterns added. Add new patterns.
* Wrapping causes the textures to be unevenly applied. Need to invert the texture transform matrix
* Fix legend
* Support for colored patterns i.e. hatches

**Comment by doutriaux1** [comment]
@sankhesh that is great! quick comment, when using pattern, fillareacolor should be ignored. When using "hatch" fillareacolors should be used to color the patterns.

**Comment by sankhesh** [comment]
@doutriaux1 

I think the white background is a subset of the current functionality. fillareacolor should be explicitly set to white (255) when using patterns.

**Comment by doutriaux1** [comment]
@sankhesh I'm just telling you how it used to work, we want to preserve backward compatibility as much as possible.

**Comment by doutriaux1** [comment]
@sankhesh it is easy enough for us to detect "pattern" and turn if back to "white" internally. Same for "hatch"

**Comment by durack1** [comment]
@doutriaux1 @sankhesh as a user, I'd certainly be happy to deal with new defaults if they really do provide progress or new functionality..

As with some of the other functions, for a change a user could be alerted with a warning message in the console for e.g. that would repeat a number of times and be turned off similar to the `cdms2.setCompressionWarnings()` function

**Comment by jypeter** [comment]
I'm not sure backward compatibility is such an issue here, when patterns have been broken for quite some time, and not overly advertised when they worked.

There can't be that many existing scripts using patterns. 

So, if breaking compatibility a  bit will make better or more useful patterns, go for it. 

As long as they look good on the screen AND in png AND pdf output... 

I've already requested transparency somewhere (maybe boxfill). But will/would transparency work for patterns? 
-- 
JYP @ SGS 2

**Comment by doutriaux1** [comment]
@durack1 you do not have years of script with hatching in them though, unlike some of our users here... So I say let's try to keep it exactly the same as much as possible.

**Comment by durack1** [comment]
@doutriaux1, my point (which mirrors @jypeter) was that if there is a better/cleaner/faster way of doing this, then go for it rather than having to tweak the settings within scripts from a few users that haven't been able to generate hatched plots for years anyway - since UV-CDAT 1.0 or similar?

If it's broke, fix it, and while you're at it make it better/cleaner/faster..

**Comment by doutriaux1** [comment]
and my point is that now we will need to
* add an attrbute for hatch color (when fillareacolor used to work)
* force the user to set colors to blank if they want patternon top of another plot (when they didn't need to)
I don't see how this is any improvement.

**Comment by durack1** [comment]
@doutriaux1 more (documented) functionality and control is good.. If I have to add another attribute/variable to my function call, but I get more control as a side-effect then it's all good for me..

In particular being able to control hatching colour would be ideal, that means you can more granular-ly control the impact of adding the hatching to your plot..

**Comment by doutriaux1** [comment]
@durack1 that exactly what the old functionality was doing and what I'm trying to get back :wink:

**Comment by sankhesh** [comment]
**UPDATE:**

Okay, I have the patterns and hatches working for isofill plots. The legend fillarea works as well.

![test_fillarea](https://cloud.githubusercontent.com/assets/936354/9480521/a5f6dfb4-4b51-11e5-9d1a-0b282cf6d709.png)


![test_fillarea_list](https://cloud.githubusercontent.com/assets/936354/9480529/b49c4bda-4b51-11e5-862e-40e40795db8e.png)


@doutriaux1 
I haven't added any additional API.
* Patterns use `fillareacolor` to render the background color. 
* Hatches use `fillareacolor` to render the hatch color.

**Comment by doutriaux1** [comment]
@sankhesh please do not use fillareacolor for pattern, leave it transparent. A tpyical use case of pattern/hatch is to plot an isofill of a variable and then plot a patterned isofill on top of it to show area of statistical significance. So you do need to see the original isofills plotted underneath it.

**Comment by doutriaux1** [comment]
@sankhesh look at @jypeter original png.

**Comment by durack1** [comment]
@sankhesh as an example I'd want to use stippling/hatching to overplot already existing colours showing regions of significance/non-signficance - so transparency is certainly important to me. For a visual example I've appended a file below:
![131014_durack wijffels2010_basinzonalchgs_ar5-final-fig3p9-colourprofile-cmyk](https://cloud.githubusercontent.com/assets/3229632/9482264/14fa1ef4-4b47-11e5-8f6e-caf974fbbcd2.jpg)

**Comment by bonfils2** [comment]
@sankhesh I love the diversity of hatches, but I agree with @doutriaux1 and @durack1 and confirm that transparency is the most crucial feature. I will exclusively use the hatches function to display regions of significance.

**Comment by sankhesh** [comment]
Sure. I'll make the pattern background transparent.

**Comment by durack1** [comment]
@sankhesh transparent by default would be awesome.. If it's not too hard, also being able to toggle that would also be useful functionality, if it's already in there and easy to expose that would be ideal..

**Comment by doutriaux1** [comment]
@durack1 i think the toggle might just add a level of complexity which can easily be replace by simply plotting the data once, once with pattern on, once w/o... The simpler the code the easier it will be to maintain!

**Comment by durack1** [comment]
@doutriaux1 my preference is to add *well documented* functionality, with intelligent defaults.. If there are options for granular control I'd like them exposed for *power* users..

**Comment by sankhesh** [comment]
@doutriaux1 @durack1 @aashish24 @bonfils2 

I added a parameter called `opacity` to the `fillarea` template. This opacity behaves differently based on the fill area `style` as follows:

* solid: Solid fill opacity
* pattern: Background fill opacity
* hatch: Foreground opacity

**Exhibit 1: Patterns with different opacity values**
![test_fillarea](https://cloud.githubusercontent.com/assets/936354/9527835/5ab1eaac-4cbf-11e5-827b-a29300390635.png)


**Exhibit 2: Hatches with different opacity values**
![test_fillarea](https://cloud.githubusercontent.com/assets/936354/9527808/28525cae-4cbf-11e5-84da-6908a2041cbf.png)

**Comment by doutriaux1** [comment]
@sankhesh so you made the opacity a list, correct? That's a great idea, I was thinking of just one value applied to all levels, but that's even better!

**Comment by durack1** [comment]
@sankhesh this is great, is it possible to also include a "stippling" option in the hatch/pattern selections - so most specifically [as above](https://github.com/UV-CDAT/uvcdat/issues/541#issuecomment-134770577)

**Comment by doutriaux1** [comment]
@durack1 what is stippling? Another pattern style?

**Comment by durack1** [comment]
@doutriaux1 @sankhesh so the opaque grey dots in the [figure above](https://github.com/UV-CDAT/uvcdat/issues/541#issuecomment-134770577)

**Comment by doutriaux1** [comment]
so like hatch10 but horizontal? @sankhesh how hard is it to add patterns? Could we somehow pass some png or vtk object instead of a pattern number and use this?

**Comment by durack1** [comment]
@doutriaux1 @sankhesh if there are a bunch of VTK options like this, it'd be great to expose them too.. The [hatch/patterns above](https://github.com/UV-CDAT/uvcdat/issues/541#issuecomment-135500323) are pretty 1980s technology in my opinion.. They look pretty klunky..

**Comment by doutriaux1** [comment]
@durack1 @sankhesh remeber we need to stay backend-agnostic as much as possible, a png would probably be what works best.

**Comment by durack1** [comment]
@doutriaux1 if we have access to more functionality/patterns/options and modernization through the VTK backend, then expose it.. There's no good reason to limit functionality in the off chance that you might want to change the backend renderer some day.. That's way too conservative.. The plotting tools and options in UV-CDAT should be up-to-date as of 2015-08-27..

In the case that a functionality wasn't available in the future, this would need to be caught with an error message.. Which would happen anyway.. And obviously other tweaks would also be needed..

My recommendation would be modernize and expose everything we have! Make it the best it can be..

**Comment by doutriaux1** [comment]
but that's exactly why vcs lived through 30years or so worth of languages and technologies :wink: you need flexibility. Every time I introduced rigidity it came back to haunt me. We can always hook up things from the backend directly if needed for "expert" users.

**Comment by doutriaux1** [comment]
vcs.backend gives you access to VTK so if you want to moveon to the VTK world purely you can still do it. But that's no longer vcs.

**Comment by durack1** [comment]
@doutriaux1 neither of the comments above are a reasonable justification to limit forward development in my opinion.. There aren't many folks who have clung to the ever reliable Windows 1995, just because it's been around 20 years..

I want to plot 21st century graphics using UV-CDAT, whether it's `VCS` or `VTK` or `matplotlib` or something else I don't really care, I just want high resolution, high production quality graphics.. And fine granular control to generate them and export them to differing formats..

**Comment by doutriaux1** [comment]
@durack1 that's what @sankhesh and I are working on, getting to you the features you want in a fashion that will not render our software obsolete. My point was exactly that vcs went through the ages because it is flexible enough to handle ever changing technologies and integrate them in a nice way.

**Comment by bonfils2** [comment]
@sankhesh : small dots would be great too!
![slide1](https://cloud.githubusercontent.com/assets/3536584/9535067/6fe40a64-4cd2-11e5-96ae-07fc31ac456b.png)

**Comment by sankhesh** [comment]
How about this? - I can add additional patterns (stipple and small dots) to the mix and get these changes merged in. Once that is done, I can look at adding the ability to pass a PNG image to the backend. I'll add that as a separate issue.

**Comment by durack1** [comment]
@sankhesh that would be a great start.. What other graphics modes/patterns/hatches are available in `VTK` that you could expose? Is there a gallery/webpage somewhere that lists these?

**Comment by sankhesh** [comment]
@durack1 The current code-base uses [vtkImageCanvasSource2D](http://www.vtk.org/doc/nightly/html/classvtkImageCanvasSource2D.html) to generate the pattern images. All the above patterns are generated using the same.
